Update to allow reinitiating pull request

marklhc opened the pull request:
This is for re-opening PR #38 so that Winnie and I can review and comment on the changes.

marklhc committed:
Update to allow reinitiating pull request

marklhc commented on R/tspa_plot.R:
```diff
@@ -0,0 +1,149 @@
+#'
+#' @param tspa_fit An object of class \code{lavaan},
+#'                 representing the output generated from `tspa()` function.
+#' @param ask Logic input. If 'True' is indicated, the user will be asked before
```
Be consistent with the R syntax in "TRUE" and "FALSE."

marklhc commented on vignettes/tspa-plot-vignette.Rmd:
````diff
@@ -0,0 +1,193 @@
+```{r load-pkg, message = FALSE}
+library(lavaan)
+library(R2spa)
+library(semdplot)
````
Please remove the dependency of `semdplot` from this branch.

marklhc commented on R/tspa_plot.R:
```diff
@@ -0,0 +1,149 @@
+#' @param ask Logic input. If 'True' is indicated, the user will be asked before
+#'            before each plot is generated. The default setting is 'False'.
+#' @param title Character. A default or user-defined name or of the title of scatterplot.
+#' @param label_x Character. A default or user-defined name of the x-aix of scatterplot.
```
What is "default" here?

marklhc commented on R/tspa_plot.R:
```diff
@@ -0,0 +1,149 @@
+                       abbreviation = TRUE,
+                       ...) {
+
+  fit_data <- parameterestimates(tspa_fit)
```
`parameterestimates` needs to be imported from `lavaan`

marklhc commented on R/tspa_plot.R:
```diff
@@ -0,0 +1,149 @@
+  latent_scores <- lavInspect(tspa_fit, what = "data")
+
+  if (is.list(latent_scores)) {
+    latent_names <- colnames(latent_scores[[1]])
```
`latent_names` was not used subsequently

marklhc commented on R/tspa_plot.R:
```diff
@@ -0,0 +1,149 @@
+    latent_iv <- c(t(na.omit(fit_data[1:(nrow(fit_data)/length(latent_scores)), ][which(fit_data$op == "~"),]["rhs"])))
+    latent_model <- list()
+
+    for (g in seq(length(df_latent_scores))) {
```
Do you mean `seq_len()`?

Gengrui-Zhang replied:
`df_latent_scores` is a list object, and using `seq_len()` will result in an error message of "argument must be coercible to non-negative integer"

marklhc replied:
You should use `seq_along()` then

marklhc commented on R/tspa_plot.R:
```diff
@@ -0,0 +1,149 @@
+                       label_y = NULL,
+                       abbreviation = TRUE,
+                       ...) {
+
```
Is there a way we can check whether the input model is a result of `tspa()`?

marklhc commented on R/tspa_plot.R:
```diff
@@ -0,0 +1,149 @@
+
+    for (g in seq(length(df_latent_scores))) {
+      for (i in seq(length(latent_dv))) {
+        latent_model[[g]] <- lm(as.numeric(t(df_latent_scores[[g]][paste0("fs_", latent_dv[i])])) ~
```
Should we show the actual relationship from the `tspa()` fit, instead of relying on `lm()`? The coefficients are already available, and may be different from the results of `lm()`. This is likely more a problem when you have multiple predictors for an outcome latent factor.

marklhc commented on R/tspa_plot.R:
```diff
@@ -0,0 +1,149 @@
+    }
+
+  } else {
+    latent_names <- colnames(latent_scores)
```
Is this part the same as the ones in the multiple-group version? Would it be more efficient write a function for single group, and then if the input is multiple group, you can just call the the single-group function multiple times?

marklhc reviewed (commented):
@Gengrui-Zhang, see my suggestions for revision. Also, please run R CMD Check to make sure your PR passes those checks. This includes functions not imported, and some other potential problems.

marklhc committed:
Update some formatting suggestions

winniewytse commented:
@marklhc and @Gengrui-Zhang, as I was unable to add comments to this pull request earlier, I added my comments under pull request #38. Would it work for you?

marklhc commented:
> @marklhc and @Gengrui-Zhang, as I was unable to add comments to this pull request earlier, I added my comments under pull request #38. Would it work for you?

I think I wasn't able to see the comments in the closed PR. Perhaps @Gengrui-Zhang can see them in his email?

Gengrui-Zhang committed:
Commit changes

Gengrui-Zhang committed:
commit change

Merge branch 'tspa-dignostic-plot' of https://github.com/Gengrui-Zhang/R2spa into tspa-dignostic-plot

# Conflicts:
#	R/tspa_plot.R

Gengrui-Zhang committed:
commit changes

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
commit

marklhc commented on R/tspa_plot_revised.R:
```diff
@@ -0,0 +1,123 @@
+tspa_plot <- function(tspa_fit,
```
Do you intend to have this "revised" file to replace the original file? Generally it is not a good idea to keep two versions of the same file using Git, as that means different file path history. Just keep one copy; when you commit a file, the history is already saved, so you don't need to have a separate file.

Gengrui-Zhang replied:
Sounds good! I deleted the old version and renamed this file to "tspa_plot.R"

marklhc commented on R/tspa_plot_revised.R:
```diff
@@ -0,0 +1,123 @@
+                      label_x = NULL,
+                      label_y = NULL,
+                      abbreviation = TRUE,
+                      fscore_type = NULL,
```
If you're allowing users to specify a type, the argument should be something like `fscore_type = c("type1", "type2")`, and then use `match.arg` inside the function. See `get_fs()` for an example.

Gengrui-Zhang replied:
Thanks for the suggestion. I made the revision accordingly.

marklhc commented on R/tspa_plot_revised.R:
```diff
@@ -0,0 +1,123 @@
+    if (is.null(fscore_type)) {
+      fscores <- fscores_est
+    } else {
+      fscores <- fscore_func
```
The variable names do not match. And if you're using conditional statements, put the computation inside the if-else, so that you don't need to do unnecessary computations. E.g.,

`if (...) {
    fscores <- lavaan::lavInspect(...)
} else {
    fscores <- ...
}

Gengrui-Zhang replied:
```
if (fscores_type == "original") {
      fscores <- lavaan::lavInspect(tspa_fit, what = "data") # factor scores from `get_fs`
    } else {
      fscores <- lavaan::lavPredict(tspa_fit) # factor scores from estimation using `lavPredict`
    }
```

Regarding "the variable names do not match", do you mean "fscores" and "fscore_type"?  Do you think adding a "s" solve this problem? Please let me know if I misunderstand anything. Thanks!

marklhc commented on R/tspa_plot_revised.R:
```diff
@@ -0,0 +1,123 @@
+  fscores_est <- lavaan::lavPredict(tspa_fit) # factor scores from estimation using `lavPredict`
+
+    # Type of scatterplot
+    if (is.null(fscore_type)) {
```
It is a strange use of `is.null`, as right now it means users can input anything other than `NULL` to trigger the alternative specification. Be more specific.

Gengrui-Zhang replied:
Solved using `fscores_type == "original"`

marklhc commented on R/tspa_plot_revised.R:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    # Helper function for scatter plot
+    # Comment: put it outside the main function
```
As mentioned, have the function separate if possible.

Gengrui-Zhang replied:
Solved.

marklhc commented on R/tspa_plot_revised.R:
```diff
@@ -0,0 +1,123 @@
+
+      plot(iv_data,
+           dv_data,
+           ylab = ifelse(is.null(label_y), paste0("fs_", dv),
```
From the documentation of `ifelse`:

> Further note that if(test) yes else no is much more efficient and often much preferable to ifelse(test, yes, no) whenever test is a simple true/false result, i.e., when length(test) == 1.

And it doesn't seem the variable `i` is defined.

Gengrui-Zhang replied:
Hi Mark, I revised this part as well. Thanks for the suggestion.

marklhc commented on R/tspa_plot_revised.R:
```diff
@@ -0,0 +1,123 @@
+           dv_data,
+           ylab = ifelse(is.null(label_y), paste0("fs_", dv),
+                         ifelse(length(label_y) > 1, label_y[i], label_y)),
+           xlab = ifelse(is.null(label_x), paste0("fs_", iv),
```
`label_y`, `label_x`, etc, should be an argument of the inner function.

Gengrui-Zhang replied:
I added those arguments to plot helper functions.

Gengrui-Zhang committed:
commit

Gengrui-Zhang committed:
update the plot function

Gengrui-Zhang committed:
update tspa plot function

Gengrui-Zhang committed:
Delete old version

Gengrui-Zhang committed:
Addressed PR #41

marklhc commented on R/tspa_plot.R:
```diff
@@ -0,0 +1,269 @@
+
+# Helper function for the scatter plot
+# Comment: put it outside the main function
+plot_scatter <- function (fscores_df, iv, dv,
```
Right now the documentation applies to `plot_scatter()`. Please keep the documentation close to the main function, and move helpers to the bottom.

marklhc commented on R/tspa_plot.R:
```diff
@@ -0,0 +1,269 @@
+#'
+#' @param tspa_fit An object of class \code{lavaan},
+#'                 representing the output generated from `tspa()` function.
+#' @param title Character. Set the name of scatter plot. The default value is "Scatterplot".
```
Avoid lines with 80+ characters

marklhc commented on R/tspa_plot.R:
```diff
@@ -0,0 +1,269 @@
+#' @param label_x Character. Set the name of the x-axis. The default value is "fs_" followed by variable names.
+#' @param label_y Character. Set the name of the y-axis. The default value is "fs_" followed by variable names.
+#' @param ... Additional arguments passed to \code{\link[graphics]{plot}}. See
+#'            \code{\link[graphics]{plot}} for a list.
```
Use `TRUE` instead of 'TRUE' for code

marklhc commented on R/tspa_plot.R:
```diff
@@ -0,0 +1,269 @@
+#'            before each plot is generated. The default setting is 'False'.
+#' @param fscore_type Character. Set the type of factor score for input.
+#'                    The default setting is using factor score from observed data (i.e., output from 'get_fs').
+#'                    If 'fscore_type = "est"', then it will use output from 'lavPredict'
```
See [here](https://roxygen2.r-lib.org/articles/markdown.html) for how to link different functions.

marklhc commented on R/tspa_plot.R:
```diff
@@ -0,0 +1,269 @@
+#' @param tspa_fit An object of class \code{lavaan},
+#'                 representing the output generated from `tspa()` function.
+#' @param title Character. Set the name of scatter plot. The default value is "Scatterplot".
+#' @param label_x Character. Set the name of the x-axis. The default value is "fs_" followed by variable names.
```
There is an extra `...`.

marklhc committed:
Reformat code and organize in correct order.

- Also resolve issues in documentation

marklhc committed:
Remove unnecessary conditions

marklhc committed:
Fix a bug in the date of vignette

marklhc committed:
Remove unnecessary dots in argument names

marklhc committed:
Further simplication, and add documentation

marklhc committed:
Fix bug in unmatched argument name

marklhc committed:
Use built-in for `ask = TRUE`

marklhc committed:
Merge branch 'devel' into tspa-dignostic-plot

marklhc commented:
@Gengrui-Zhang, thanks for working on the function. I tried merging in `devel`, and I think the latest updates in the `tspa()` function causes some errors in the vignette. Could you check and update the vignette accordingly? Thanks.

Gengrui-Zhang committed:
revised the vignette

marklhc committed:
Make format consistent

marklhc committed:
Add dependency
